### PR TITLE
[Google Drive] Connector implementation

### PR DESCRIPTION
**Pull request by jedrazb**
## Closes https://github.com/elastic/enterprise-search-team/issues/4992

Implementation of the Google Drive connector based on [Google Drive API v3](https://developers.google.com/drive/api/reference/rest/v3). 

#### Authentication

The connector is authenticated as a service account.
- [Service account overview](https://cloud.google.com/iam/docs/service-account-overview)

#### Ingestion and content extraction

- Connector extracts and ingests same fields as Workplace Search Connector
- Support for downloading Google Suite documents and stored files and folders
- Content extraction is supported
- It uses `aiogoogle` library under the hood for async API calls

#### Testing
- Unit tests, 98% coverage
- Functional test added 
- Manual tests on different sizes of data corpus (up to 85k docs) with:
  - personal account
  - domain (organizational) account Screenshot 2023-06-28 at 11 17 28

- Local run of perf8 is happy ✅ 
Screenshot 2023-07-04 at 16 36 02

#### Documentation

TODO

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

### CODE REVIEW

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [inline connectors/sources/google_drive.py:27]: Idk if it's worth to put to constants - it seems to be used in a single place, so can either be moved to the only class using it or even inlined IMO

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [inline connectors/sources/google_drive.py:160]: Avoid using `asyncio.sleep` and use `CancellableSleeps` from `utils.py`.

Problem with `asyncio.sleep` is that it will keep the process running if you want to terminate it, for example you do `await asyncio.sleep(60)` and then CTRL+C the host - it'll have to wait until `asyncio.sleep` wakes up, which is 60 seconds - too much. We want to be able to cancel these sleeps as well.

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [inline connectors/sources/google_drive.py:232]: Can you also validate the private_key? Will `get_pem_format` raise an error if private key is invalid?

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [inline connectors/sources/google_drive.py:243]: I think a logger debug line is worth to add here, that will indicate that the method is skipped for FTEST

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [inline connectors/sources/google_drive.py:271]: Do you think it makes sense to make actually rich client with these methods? E.g. your client will have methods `get_drives`, `retrieve_all_drives` and such and the connector class itself will contain only the logic to call these methods and create view models, etc

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [inline connectors/sources/google_drive.py:294]:
```suggestion
        async for page in self.get_drives():
```

It's probably page, right?

Chunk is more about files IMO

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [inline connectors/sources/google_drive.py:337]: Can we live without this cache? How much memory will it occupy for large drives?

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [inline connectors/sources/google_drive.py:405]:
```suggestion
        logger.debug(f"Downloaded {blob_name} with size of {blob_size} bytes ")
```

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [inline connectors/sources/google_drive.py:448]: Is it possible to not even download the file if it's too big?

**Review by artem-shelkovnikov — COMMENTED** [review]
Great start with a connector! I've left some comments, happy to also jump on a call to discuss any time.

File: connectors/sources/google_drive.py
Comment (abhishekjoshi-crest) [inline connectors/sources/google_drive.py:115]: Can we add separate method to handle paginated calls?

**Review by abhishekjoshi-crest — COMMENTED** [review]
Great work with path lookup implementation. Just minor suggestion regarding new logger and method separation.

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:367]: It takes ~30s for 15k folders in multiple drives. Yup, it's blocking as we need to build the path lookup before processing the files.

Comment (artem-shelkovnikov) [reply connectors/sources/google_drive.py:367]: You can unblock it with sleeping for 0 seconds every now and then - it'll yield to another task and return back eventually

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [reply connectors/sources/google_drive.py:448]: Good point, thanks for the explanation! Can you add a code comment about it?

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:337]: Takes `0.5MB` of mem for a setup with 15k folders, multiple drives and 85k files in total.

This cache is needed if we want to have the feature parity with Workplace Search. There is no other way to determine the path on drive. See corresponding implementation in: https://github.com/elastic/ent-search/blob/main/connectors/lib/connectors/content_sources/google_drive/extractor.rb#L61 

One reason for keeping this would be the ability to perform search for a given subpath. On the other hand we could just keep track of `drive_name` where a given file exists (that would be much faster as the lookup will just contain drives).

Comment (seanstory) [reply connectors/sources/google_drive.py:337]: The reason Workplace Search built up these paths was to support filtering rules, IIRC. As is, this would be useful with Basic Sync Rules to help accomplish tasks like "filter all documents in the tree under `/unwanted-dir/`". However, Workplace Search didn't have Advanced Sync Rules, and that's really where we'd want to accomplish this kind of task. If we can limit what trees we access at fetch time, we may not need to build up these paths like this. Might be worth checking with Dana and her plans for Advanced Sync Rules with GDrive.

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:144]: The `def _retry(retry_counter, log_message)` performs a condition check if current `retry_counter` is bigger than max retries and if true it would re-raise the exception. 

That being said, the `@retryable` decorator seems to be cleaner approach and I will amend the code to use it.

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:250]: It was needed because `api_call()` was returning async generator for both paginated and non-paginated calls, await anext was get the first (and single) object from the async iterator. 

Following suggestions from other comments I'm splitting the api_call function so that the paginated one will be returning async generator whereas the non-paginated func will be returning awaitable result.

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:337]: From what I understand we can pass a query `q` to as parameter to API call to filter results at fetch time. Here is the documentation for reference:
- https://developers.google.com/drive/api/guides/ref-search-terms 

From what I understand it only supports filtering for shared drive membership, it doesn't support path-based filtering. I will confirm with Dana the right way forward.

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:114]: because the for e.g. `files/list` endpoint we need to access `drive_client.files.list` method object. 

Since `resource` and `method` are api_call params this is the way to programatically access the object. Note that (resource, method) are different for different api calls, they can be:
- (`files`, `list`)
- (`files`, `get`)
- (`drives`, `list`)
- ...

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:101]: I tried to have a per-client instance of aiogoogle, but I had some warnings/errors about unclosed client sessions. Also this pattern is suggested on [aiogoogle docs](https://aiogoogle.readthedocs.io/en/latest/)

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:448]: Also worth mentioning is that Google Export endpoint will return `403` codes for exports larger than >5MB (that's also good in preventing large downloads), see https://github.com/elastic/enterprise-search-team/issues/1374

File: connectors/sources/google_drive.py
Comment (danajuratoni) [reply connectors/sources/google_drive.py:337]: Customers have often mentioned that for file based systems, it's extremely valuable to restrict the ingested scope by sharing only specific folders out of an entire GDrive tree structure or the opposite. 

Basic rules, especially with the current first match logic, are "born" to be used for this type of filtering so it would be awesome if we could leverage this. If the performance impact is not worth the cost, I'm open to give Advanced Sync Rules this responsibility completely.

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:106]: Thank you for spotting this. 👍  I will use the new logging implementation.

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:196]: Yup just realized we can use `connectors.json` to pass the default configuration for ftests. Agreed that it's pretty much useless 👍

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:232]: `get_pem_format` is actually not needed - I copy pasted this from google cloud storage client. It was used to handle default credentials.

I realised we can use `connecotrs.json` for providing credentials for ftests. 

`get_pem_format` just formats the private key into single line format. I would expect that `aiogoogle` or `google api` will error out if private key is invalid. User needs to only copy-paste the generated json here.

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:400]: It's fast, it doesn't read the file to determine its size. It seems to be the preferred way for checking file size.

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:350]: I timed the blocking part of `resolve_paths()` for 15k folders, with:
```
--- a/connectors/sources/google_drive.py
+++ b/connectors/sources/google_drive.py
@@ -326,6 +326,8 @@ class GoogleDriveDataSource(BaseDataSource):
         folders = await self.retrieve_all_folders()
         drives = await self.retrieve_all_drives()
 
+        import time
+        start_time = time.time()
         # for paths let's treat drives as top level folders
         for id, drive_name in drives.items():
             folders[id] = {"name": drive_name, "parents": []}
@@ -349,6 +351,10 @@ class GoogleDriveDataSource(BaseDataSource):
 
             folder["path"] = "/".join(path)  # Join path elements with '/'
 
+        end_time = time.time()
+
+        self._logger.info(f"Resolving paths took: {end_time - start_time}s")
+
         return folders

```

Here is the result:
```
[...] Resolving paths took: 0.03356814384460449s
```

Do you think I should still introduce periodic `asyncio.sleep(0)` calls?

Comment (artem-shelkovnikov) [reply connectors/sources/google_drive.py:350]: I think it's good, thanks for checking :)

File: connectors/sources/google_drive.py
Comment (wangch079) [inline connectors/sources/google_drive.py:391]: Personally, I'm not a fan of temp files. In the [official guide](https://developers.google.com/drive/api/guides/manage-downloads#python), there is an example of using IO streaming. Have you considered that approach?

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:371]: Thank you for your feedback, the parents can also point to drives (from my testing).  This approach is consistent with what we have in `ent-search` ([link](https://github.com/elastic/ent-search/blob/main/connectors/lib/connectors/content_sources/google_drive/extractor.rb#L66))

Here is a result from internal testing:
Screenshot 2023-07-05 at 10 04 19

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:502]: The framework pops `_id` from referenced blob and sets  its value to `id` key, this is executed before calling `get_content`. I don't have access to `_id` key at this step.

https://github.com/elastic/connectors-python/blob/main/connectors/es/sink.py#L392

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:450]: The framework pops `_id` from referenced blob and sets  its value to `id` key, this is executed before calling `get_content`. I don't have access to `_id` key at this step.

https://github.com/elastic/connectors-python/blob/main/connectors/es/sink.py#L392

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:383]: Good catch, thank you! Will add the improvement :)

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:627]: We are expecting the key `size` to be present and sometime it point to `None` value, afaik the `get()` method with default value `0` will still return `None` for keys that are present in the dict.

I did a small test:

```
>>> test_blob
{'size': None}
>>> size = test_blob.get('size', 0)
>>> print(size)
None
```

Comment (wangch079) [reply connectors/sources/google_drive.py:631]: That's a valid point. Maybe:
```python
size = test_blob.get('size') or 0
```

Comment (jedrazb) [reply connectors/sources/google_drive.py:631]: Yeah that's neat! Will include the suggestion

File: connectors/sources/google_drive.py
Comment (wangch079) [inline connectors/sources/google_drive.py:521]: This will actually be removed here:
https://github.com/elastic/connectors-python/blob/7fbb60ea586c95af4419659b8c25e28d5d022ac2/connectors/es/sink.py#L327-L328

I see the same pattern in other source implementations. Is it intended? cc @elastic/ingestion-team

File: connectors/sources/google_drive.py
Comment (wangch079) [reply connectors/sources/google_drive.py:435]: Not a blocker. We can revisit later.

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [reply connectors/sources/google_drive.py:506]: I think it's some legacy logic. I'm honestly not a big fan of the fact that once you `yield` something, the object is modified. Let's address this at some point in future

File: connectors/sources/google_drive.py
Comment (artem-shelkovnikov) [reply connectors/sources/google_drive.py:521]: Yeah afaik it's intended - @tarekziade originally implemented it this way. 

We pop `ids` always because `id` is not a part of the document - it's a part of document metadata. Why `_timestamp` is popped I don't remember. Let's create an issue to address it?

**Review by artem-shelkovnikov — APPROVED** [review]

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:360]: We decided to keep this logic, the mem footprint is minimal even for large drives (85k files). The blocking part of the function for large drives takes `~0.3s` and this is acceptable.

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:506]: https://github.com/elastic/enterprise-search-team/issues/5095

File: connectors/sources/google_drive.py
Comment (jedrazb) [reply connectors/sources/google_drive.py:521]: https://github.com/elastic/enterprise-search-team/issues/5095

Comment (wangch079) [reply connectors/sources/google_drive.py:521]: What I meant, it's not necessary to include `_id` and `_timestamp` here. It will eventually be removed from here and not used at all:
https://github.com/elastic/connectors-python/blob/7fbb60ea586c95af4419659b8c25e28d5d022ac2/connectors/es/sink.py#L327-L328

For the `lazy_download` function, it should be enough to just return a document with key `_attachment`.

But this should not block this PR.

**Review by wangch079 — APPROVED** [review]
Great work Jedr

**Pull request merged**
